### DISCUSSION TITLE
src/models下自定义model里使用了useEffect来掉接口，这种情况下如何保证wrappers/auth的鉴权处理？

### DISCUSSION THREAD

**Discussion by LXY02** [umijs/umi#11824] [Q&A] [answered]
如题，routes里配置了`{wrappers: [
      '@/wrappers/auth'
    ]}`, 打算在auth文件里调鉴权接口来判断用户是否有权限访问页面。但是例如src/models/useTestModel.ts，这个useTestModel里使用了例如`useEffect(() => {  fetchApi() // 调一些业务接口 }, [])`, 观察发现，这个useTestModel里的useEffect执行时机比auth里的useEffect要早。这种情况下如何保障可以先执行鉴权接口？我不希望把useTestModel里useEffect挪到实际使用页面里去，因为那样要写很多遍。

**Comment by fz6m** [comment, marked as answer]
在 umi 里的 model 都是全局性质的，也就是说这里面的东西和副作用在应用启动一开始就会执行了。

如果要依靠 umi 内置的功能，比较推荐的做法是在 `src/app.ts` 的 `getInitialState` 里去获取用户的初始信息、权限等状态，之后就可以从 `useModel('@@initialState')` 里获取，这样就可以在 wrapper / layout 之类的 HOC 里去拦截了，参考 [文档](https://umijs.org/docs/max/data-flow#%E5%85%A8%E5%B1%80%E5%88%9D%E5%A7%8B%E7%8A%B6%E6%80%81) 。

比较现代的做法是使用一种全局状态管理库，比如 [jotai](https://github.com/pmndrs/jotai) 或 [valtio](https://umijs.org/docs/max/valtio) ，然后在应用 root 或者最上层 layout 等位置里 `useEffect` 调用获取信息的接口，拿到信息后设定到全局状态管理库里，之后就可以在任何想要的位置获取了，在没获取到之前展示全局 loading 即可。

**Comment by zhaohongyan** [comment]
如何在登录之后再去设置全局信息？
